### PR TITLE
decaying connmgr tags for message delivery

### LINKED ISSUE #306: Implement decaying connection manager tags for first time publishers
This will ensure that recent publishers and gossipers remain connected even at the face of a connection flood and subsequent connection manager action.

Requirements:
- [ ] https://github.com/libp2p/go-libp2p-core/pull/104
- [ ] https://github.com/libp2p/go-libp2p-connmgr/pull/61

### PR DESCRIPTION
**Pull request by yusefnapora**
This adds decaying connection manager tags that give a boost to peers that deliver messages. closes #306 

I was hoping to finish and fix the test over the weekend, but didn't get a chance. It seems to basically work, but some of the honest peers aren't connected at the end of the test. 

I'll sort it out in the morning, just wanted to get the branch up in case @vyzo wants to look through before then.

### CODE REVIEW

File: gossipsub.go
Comment (vyzo) [inline gossipsub.go]: shouldn't this live inside the `tagTracer`?

File: gossipsub.go
Comment (vyzo) [inline gossipsub.go]: this should probably go inside the `tagTracer`

File: tag_tracer.go
Comment (vyzo) [inline tag_tracer.go]: let's make this a little longer if it is to decrement by integer leaps.

Comment (vyzo) [reply tag_tracer.go]: may be `10*time.Minute`?

File: tag_tracer.go
Comment (vyzo) [inline tag_tracer.go]: I think we need a nil check (probably at the beginning of the function) for this, we might panic if the connmgr doesn't support decaying tags.

**Comment by yusefnapora** [comment]
@vyzo the tag tracer is still missing "near first" message deliveries. Can you think of a good way to do this without duplicating the delivery record tracking from the score tracer?

**Comment by vyzo** [comment]
I think we can get away very simply by starting tracking in `Validate` and stopping all tracking when the message is `Delivered` or `Rejected`. The `Duplicate` should only fire while still validating.
So basically we can have a lightweight version of what we have in the score tracer.

**Comment by yusefnapora** [comment]
I squashed a few things and marked this as "ready for review", but we still need to wait until the dependency PRs are merged.

File: tag_tracer.go
Comment (vyzo) [inline tag_tracer.go]: We might want to nil the `drec.peers` map before we return the near first peers, so as to avoid holding onto memory unnecessarily. It shouldn't be used again after the delivery status has been marked as `deliveryValid`.

**Review by vyzo — COMMENTED** [review]
this looks good, just a comment.

Note: we'll want to `go mod tidy` this when it's time to merge.

File: tag_tracer.go
Comment (vyzo) [inline tag_tracer.go]: actually, there are some cases where the action (marking as invalid) is incorrect; if the rejection is because of invalid or missing signature or validation queue full we shouldn't clear the delivery record. This is because the rejection happens without entering the validation pipeline and we can end up rejecting a message that is still validating.

So let's add a check here for the reason and short-circuit if it is one of the above; see also the handling in score.go

Comment (vyzo) [reply tag_tracer.go]: Might even make sense to check explicitly for one of validation throttled, validation ignored, validation rejected as these are the only ones that indicate a message that was in the validation queue.
Then we can set the delivery status to a known value (not necessarily invalid).

File: tag_tracer.go
Comment (raulk) [inline tag_tracer.go]: I'm moving the [preset decay and bump functions](https://github.com/libp2p/go-libp2p-connmgr/pull/61/files#diff-7167ce1adea9dc2edd55d7253ca2d1f5) from `go-libp2p-connmgr` into `go-libp2p-core` so that you can reuse them here without introducing an additional module dependency.

What you've implemented here is basically the `FixedDecay` decay function, and the `SumBounded` bump function.

Comment (raulk) [reply tag_tracer.go]: Done. Mind adopting the preset functions here? https://github.com/libp2p/go-libp2p-core/pull/104/commits/d3277047d6ca1eae46b03c4816ca00e696b114e8

File: tag_tracer_test.go
Comment (raulk) [inline tag_tracer_test.go]: import order

File: gossipsub_connmgr_test.go
Comment (raulk) [inline gossipsub_connmgr_test.go]: nit: import order.

File: tag_tracer.go
Comment (raulk) [inline tag_tracer.go:75]: Do we always want to use the default? shouldn't we use the message ID function that the user has set?

File: tag_tracer.go
Comment (raulk) [inline tag_tracer.go]: Shouldn't we have a function for erasing a decaying tag explicitly?

File: tag_tracer.go
Comment (yusefnapora) [reply tag_tracer.go:75]: We set it to the router's `msgID` func in the `Start` method. Although now that I removed the background process we don't really need a `Start` method and could just put it in the constructor instead.

Comment (vyzo) [reply tag_tracer.go:75]: The actual MsgID function is not known when it is constructed, it might be changed by the options.
So keep the Start method :)

Comment (yusefnapora) [reply tag_tracer.go:75]: vyzo reminded me that we should keep it in `Start`, since the message id is set in an option that might be applied after the `tagTracer` is created.

File: tag_tracer.go
Comment (vyzo) [reply tag_tracer.go]: We don't actually use that though. When would we want to remove a tag? Perhaps when we have a rejection with validationFailed?

File: tag_tracer.go
Comment (vyzo) [inline tag_tracer.go]: that comment is not technically accurate and is confusing; the duplicate might arrive after we have Delivered/Rejected the message.

File: tag_tracer.go
Comment (vyzo) [reply tag_tracer.go]: Actually there is a case we do want to delete the tag when we remove it: imagine Joining, Leaving, and then Joining again a topic; that would lead to an error when we try to recreate the tag.
So yeah, you are right, we have to delete it.

Comment (raulk) [reply tag_tracer.go]: https://github.com/libp2p/go-libp2p-core/pull/151 and https://github.com/libp2p/go-libp2p-connmgr/pull/72 are merged.

Will wait til you integrate those features here by importing master. Will make a release once you give me the thumbs up.

**Comment by yusefnapora** [comment]
hmm, I added another 5 minutes to the travis timeout, but we're still hitting it. I can't imagine the tests in this branch take an extra 5m to run... I wonder if I could be deadlocking somewhere?

**Comment by yusefnapora** [comment]
okay, I finally wised up and figured out how to enable the race detector by default when I run tests in IntelliJ. With it enable, we hit the max goroutine limit in `TestGossipsubConnTagMessageDeliveries`. But reducing the number of peers in the test seems to make it happy.

**Comment by vyzo** [comment]
we should be getting output in that case.

**Comment by yusefnapora** [comment]
hmm, that's true. It did also take a few minutes on my laptop before it failed, so it might actually be taking more than 5m on travis

**Comment by yusefnapora** [comment]
I got the end-to-end test passing by lowering the number of test peers again, but now the unit test is failing, and it's a little confusing.

```
=== RUN   TestTagTracerDeliveryTags
    TestTagTracerDeliveryTags: tag_tracer_test.go:138: expected delivery tag value = 10, got 11
    TestTagTracerDeliveryTags: tag_tracer_test.go:145: expected delivery tag value = 0, got 1
```

This is just after adding 50 minutes to the fake clock, and checking that the tag value has decreased.

On my laptop, it gets the expected value, but it always seems to be off by one on Travis. The reason I'm confused is that the result should only depend on the fake clock we pass into the decayer, not the real clock. I tried sleeping the real clock for a second after bumping the fake one, to give the Decayer's processing routine time to run, but it still fails.

I'm going to try increasing the sleep time to five seconds, but I'll probably be asleep myself by the time it's done running. Anyway, Vyzo, perhaps you could take a look if it's still failing.

File: .travis.yml
Comment (vyzo) [inline .travis.yml]: let's restore this to 30m, it doesn't seem to change anything.

File: go.mod
Comment (vyzo) [inline go.mod]: let's not forget to `go mod tidy`.

**Review by vyzo — COMMENTED** [review]
Hard to tell what's happening with the tests on travis, code looks good to me.

File: tag_tracer.go
Comment (vyzo) [inline tag_tracer.go]: returning just nil will do.

**Review by vyzo — APPROVED** [review]
at last, green checkmaks!

**Pull request merged**
